unyt/dask_array.py

```diff
@@ -0,0 +1,620 @@
+# conversion decorator.
+_unyt_funcs_to_track = ["to", "in_units", "in_cgs", "in_base", "in_mks"]
+
+
+def _track_conversion(unyt_func_name, current_unyt_dask):
+    """
+    a decorator to use with unyt functions that convert units
+
+    Parameters
+    ----------
+
+    unyt_func_name: str
+        the name of the function to call from the sidecar _unyt_quantity. Must be
+        an attribute of unyt.unyt_quantity.
+    current_unyt_dask: unyt_dask_array
+        the current instance of a unyt_dask_array
+
+
+    Returns a new unyt_dask_array instance with appropriate units
+    """
+
+    def wrapper(*args, **kwargs):
+
+        # current value of sidecar quantity
+        init_val = current_unyt_dask._unyt_array.value
+
+        # get the unyt function handle and call it
+        the_func = getattr(current_unyt_dask._unyt_array, unyt_func_name)
+        new_unyt_quantity = the_func(*args, **kwargs)
+
+        # calculate the conversion factor and pull out the new name and units
+        # might be able to use _get_conversion_factor here too...
+        factor = new_unyt_quantity.value / init_val
+        units = new_unyt_quantity.units
+
+        # apply the factor, return new
+        new_obj = unyt_from_dask(current_unyt_dask * factor, units)
+
+        return new_obj
+
+    # functools wrap fails here because unyt_func_name is a string, copy manually:
+    wrapper.__doc__ = getattr(current_unyt_dask._unyt_array, unyt_func_name).__doc__
+    return wrapper
+
+
+# helper sanitizing functions for handling ufunc, array operations
+def _extract_unyt(obj):
+    # returns the hidden _unyt_quantity if it exists in obj, otherwise return obj
+    if _is_iterable(obj):
+        return [_extract_unyt(ob) for ob in obj]
+    return obj._unyt_array if isinstance(obj, unyt_dask_array) else obj
+
+
+def _extract_dask(obj):
+    # returns a plain dask array if the obj is a unyt_dask_array, otherwise return obj
+    if _is_iterable(obj):
+        return [_extract_dask(ob) for ob in obj]
+    return obj.to_dask() if isinstance(obj, unyt_dask_array) else obj
+
+
+def _extract_unyt_val(obj):
+    # returns the value of a unyt_quantity if obj is a unyt_quantity
+    if _is_iterable(obj):
+        return [_extract_unyt_val(ob) for ob in obj]
+    return obj.to_value() if isinstance(obj, ua.unyt_quantity) else obj
+
+
+def _check_unyt_inputs(ui_0, ui_1):
+    # returns True if the args need to be converted to the same unit, based on
+    # if the units have the same dimensions but different units.
+
+    if hasattr(ui_0, "units") and hasattr(ui_1, "units"):
+        un_0, un_1 = ui_0.units, ui_1.units
+        return un_0 != un_1 and un_0.dimensions == un_1.dimensions
+    return False
+
+
+def _is_iterable(obj):
+    return type(obj) is tuple or type(obj) is list
+
+
+def _sanitize_unit_args(*input):
+    # returns sanitized inputs and unyt_inputs for calling the ufunc
+    # what if i is a list or tuple????? np.concatenate([x,x2]) is one arg...
```

Comment:
reading back through my changes... will look into this comment I left myself...

Reply:
ok then don't hesitate to ping me when it's ready. I can review this later this week, I think